### ISSUE TITLE
Don't allow to explicitly set docker_container.command to nil

### ISSUE THREAD

**Issue by jperville** [someara/chef-docker#517]
Don't allow to explicitly set `docker_container.command` to nil, this results in redeploy of existing container. For some containers this also results in the container crashing.

Symptoms:
```
    * docker_container[nginx-proxy] action run
      - stopping nginx-proxy 
      - deleting nginx-proxy
      - update nginx-proxy
      -   set command to nil (was "forego start -r")  # <--------------------- BUG
      - starting nginx-proxy
```

The sample scenario is when a wrapper cookbook allows the user to set an attribute that will be passed as docker_container.command (to give an example, the user may want to override the "command" to '/sbin/my_init --enable-insecure-key' while in development mode). When the attribute is not set, we should just use the default command to boot the container.

Right now (as of version 2.1.0), the docker_container.command is expected to be a `ShellCommand`, a type which is declared like this:

```ruby
      ShellCommand = property_type(
        is: [String, nil],   # <----- nil here means that the attribute can be explicitly set to nil
        coerce: proc { |v| coerce_shell_command(v) }
      ) unless defined?(ShellCommand)
```

When my wrapper attribute cookbook is not set, it results in my container being converged with `command nil`. In previous versions, it would just be ignored and the default command used instead.

Now, it really sets the command to nil, which results in my container crashing on start:

```
root@staging3-compute1:~# docker ps -n=10 | grep nginx-proxy
446637439d0d        XXXXXXXXXXX/nginx-proxy:latest             "/app/docker-entrypoi"   2 minutes ago       Restarting (0) 40 seconds ago   192.168.0.25:80->80/tcp, 127.0.0.1:7776->443/tcp     nginx-proxy
```

As a workaround, I had to modify my wrapper cookbook to NOT pass the command if the value is nil.

```
docker_container 'mycontainer' do
   ...
   command node['mycookbook']['docker']['command'] unless node['mycookbook']['docker']['command'].nil?
end
```

This is an ugly workaround since (IMO) there is no valid use case for allowing the docker container command to be passed to the API as nil instead of being ignored as before.

**Comment by chasebolt** [comment]
if i'm understanding this correctly, this relates to #514. fixing that should resolve this issue as well. again, this is from the merging of the resources (docker create + Dockerfile).

**Comment by jperville** [comment]
Actually, it is totally unrelated to #514.

The reason of this issue is that recent chef versions (12.5+) will set an attribute value to `nil` when the resource is declared with an explicit `nil` in the list of valid classes (eg. `property :myprop, is: [String, nil]`). With earlier chef versions, the value would be ignored, as the property assignment to nil was treated like a getter.

I learn about this because the chef_compat now emits a deprecation warning in this case : https://github.com/chef-cookbooks/compat_resource/blob/master/files/lib/chef_compat/copied_from_chef/chef/property.rb#L258

In this cookbook, since the `command` property is expected to be `[String, nil]` then it means that setting:

```
docker_container 'mycontainer' do
  ...
  command nil
end
```
In my case, the `nil` value is coming from evaluating an attribute which happens to have no default value (eg. `node['mycookbook']['docker']['command'] = nil`).

In earlier version of this cookbook (such as 1.0.x, which did not depend on compat_resource to backport the chef 12.5 behavior) the command assignment to nil was ignored and the default command from the Dockerfile was used.

**Comment by chasebolt** [comment]
thanks for clearing that up, that makes sense now.

not sure why you would be hitting the deprecation warning though. the property explicitly accepts nil and it shouldnt be falling into this conditional.
https://github.com/chef-cookbooks/compat_resource/blob/master/files/lib/chef_compat/copied_from_chef/chef/property.rb#L481-L493

@jkeiser, bringing you in since you wrote the property_types. is nil required in this property_type or can it be removed to match what was accepted in 1.x?

**Comment by jperville** [comment]
The property accepts nil and that's why it results in a container that chain-restarts for me (the command is overwritten by the nil if I specify the properly). If the property did not accept nil, then yes I would have fall on the deprecation (I did not trigger it, but I know its existence from investigating earlier issues).

**Comment by someara** [comment]
released in 2.1.4
